### PR TITLE
Fundamental and essential matrix 8-point algorithm

### PR DESCRIPTION
**Pull request by ahojnnes**
Just here to remind me to finish this at some point in the near future.

UPD (@soupault): closes also https://github.com/scikit-image/scikit-image/issues/2318.

### CODE REVIEW

**Comment by stefanv** [comment]
Yes!! This will be great to have!

**Comment by ahojnnes** [comment]
@scikit-image/core This should be ready to review. I plan to add more 3D functionality sometime soon but I would like to get this PR approved before adding more things. This should be feature-complete, includes unit tests and an example for the gallery. As I add more functionality, I will expand on the examples as well.

File: doc/examples/transform/plot_fundamental_matrix.py [hunk 1]
```diff
@@ -0,0 +1,57 @@
...
+
+"""
+from skimage import data
+from skimage.color import rgb2gray
+from skimage.feature import match_descriptors, ORB, plot_matches
+from skimage.measure import ransac
+from skimage.transform import FundamentalMatrixTransform
+import matplotlib.pyplot as plt
+
+img1, img2 = map(rgb2gray, data.middlebury2014_motorcycle())
```
Comment (jni) [inline doc/examples/transform/plot_fundamental_matrix.py]: I'd rename this data element, it's too wordy imho. How about `stereo_motorcycle`? The citation can live in the docstring.

File: skimage/transform/_geometric.py [hunk 2]
```diff
@@ -208,8 +209,275 @@ def __add__(self, other):
...
+        E_normalized, src_matrix, dst_matrix = \
+            self._setup_constraint_matrix(src, dst)
+
+        # Enforcing the internal constraint that two singular values must be
+        # equal and one must be zero.
+        U, S, V = np.linalg.svd(E_normalized)
+        S[0] = (S[0] + S[1]) / 2.0
+        S[1] = S[0]
+        S[2] = 0
+        E = np.dot(U, np.dot(np.diag(S), V))
```
Comment (jni) [inline skimage/transform/_geometric.py]: It boggles my mind that this doesn't bother you. =P

File: skimage/transform/_warps.py [hunk 3]
```diff
@@ -402,3 +412,399 @@ def swirl(image, center=None, strength=1, radius=100, rotation=0,
     return warp(image, _swirl_mapping, map_args=warp_args,
                 output_shape=output_shape, order=order, mode=mode, cval=cval,
                 clip=clip, preserve_range=preserve_range)
+
+
+def _stackcopy(a, b):
+    """Copy b into each color layer of a, such that::
```
Comment (jni) [inline skimage/transform/_warps.py]: For future PRs, I would recommend that we only do big function moves when we are making no other changes, because GitHub doesn't make it easy to see whether anything's changed during the move.

File: skimage/transform/tests/test_geometric.py [hunk 4]
```diff
@@ -203,6 +194,67 @@ def test_piecewise_affine():
     assert_almost_equal(tform.inverse(DST), SRC)
 
 
+def test_fundamental_matrix_estimation():
```
Comment (jni) [inline skimage/transform/tests/test_geometric.py]: Can you add a note about where these coordinates come from?

File: skimage/transform/_geometric.py [hunk 5]
```diff
@@ -158,7 +160,7 @@ def __call__(self, coords):
         Returns
         -------
         coords : (N, 2) array
-            Transformed coordinates.
+            Destination coordinates.
```
Comment (ahojnnes) [inline skimage/transform/_geometric.py]: Yes.

File: skimage/transform/_geometric.py [hunk 6]
```diff
@@ -208,8 +209,275 @@ def __add__(self, other):
...
+        A[:, 0] *= dst[:, 0]
+        A[:, 1] *= dst[:, 0]
+        A[:, 2] *= dst[:, 0]
+        A[:, 3:5] = src
+        A[:, 5] = 1
+        A[:, 3] *= dst[:, 1]
+        A[:, 4] *= dst[:, 1]
+        A[:, 5] *= dst[:, 1]
+        A[:, 6:8] = src
+        A[:, 8] = 1
```
Comment (ahojnnes) [inline skimage/transform/_geometric.py]: Your code is not doing the same thing, but I simplified anyway.

File: skimage/transform/_geometric.py [hunk 7]
```diff
@@ -208,8 +209,275 @@ def __add__(self, other):
...
+
+        # Enforcing the internal constraint that two singular values must non-
+        # zero and one must be zero.
+        U, S, V = np.linalg.svd(F_normalized)
+        S[2] = 0
+        F = np.dot(U, np.dot(np.diag(S), V))
+
+        self.params = np.dot(dst_matrix.T, np.dot(F, src_matrix))
+
+        return True
```
Comment (ahojnnes) [inline skimage/transform/_geometric.py]: Yes :-)

File: skimage/transform/_warps.py (hunk 3 above)
Comment (ahojnnes) [inline skimage/transform/_warps.py]: Agreed, but you have to believe me that I simply copy&pasted the code.

File: skimage/data/__init__.py [hunk 8]
```diff
@@ -321,3 +322,51 @@ def rocket():
...
     return load("rocket.jpg")
+
+
+def middlebury2014_motorcycle():
+    """Rectified stereo image pair.
+
+    The two images are rectified such that every pixel in the left image has its
+    corresponding pixel on the same scanline in the right image. That means that
+    both images are warped such that they have the same orientation but a
+    baseline in the x-direction.
```
Comment (ahojnnes) [inline skimage/data/__init__.py]: Fixed.

File: skimage/data/__init__.py [hunk 9]
```diff
@@ -321,3 +322,51 @@ def rocket():
...
+    -------
+    img_left : (512, 345, 3) uint8 ndarray
+        Left stereo image.
+    img_right : (512, 345, 3) uint8 ndarray
+        Right stereo image.
+
+    Notes
+    -----
+    The original resolution images, images with different exposure and lighting,
+    and ground-truth depth maps can be found at the Middlebury website [2]_.
```
Comment (ahojnnes) [inline skimage/data/__init__.py]: Done and example expanded.

File: skimage/transform/_geometric.py [hunk 10]
```diff
@@ -208,8 +209,275 @@ def __add__(self, other):
...
+            Source coordinates.
+
+        Returns
+        -------
+        coords : (N, 3) array
+            Epipolar lines in the destination image.
+
+        """
+        coords_homogeneous = np.column_stack([coords, np.ones(coords.shape[0])])
+        return np.dot(coords.homogeneous, self.params.T)
```
Comment (ahojnnes) [inline skimage/transform/_geometric.py]: Fixed and unit test added.

File: skimage/transform/_geometric.py [hunk 11]
```diff
@@ -208,8 +209,275 @@ def __add__(self, other):
...
+            Destination coordinates.
+
+        Returns
+        -------
+        coords : (N, 3) array
+            Epipolar lines in the source image.
+
+        """
+        coords_homogeneous = np.column_stack([coords, np.ones(coords.shape[0])])
+        return np.dot(coords.homogeneous, self.params)
```
Comment (ahojnnes) [inline skimage/transform/_geometric.py]: As above... :-)

File: skimage/transform/_geometric.py [hunk 12]
```diff
@@ -208,8 +209,275 @@ def __add__(self, other):
...
+
+    def _setup_constraint_matrix(self, src, dst):
```
Comment (ahojnnes) [inline skimage/transform/_geometric.py]: Done.

File: skimage/transform/_geometric.py [hunk 13]
```diff
@@ -208,8 +209,275 @@ def __add__(self, other):
...
+        assert src.shape == dst.shape
+        assert src.shape[0] >= 8
+
+        # Center and normalize image points for better numerical stability.
+        try:
+            src_matrix, src = _center_and_normalize_points(src)
+            dst_matrix, dst = _center_and_normalize_points(dst)
+        except ZeroDivisionError:
+            self.params = np.nan * np.empty((3, 3))
```
Comment (ahojnnes) [inline skimage/transform/_geometric.py]: Fixed.

File: skimage/transform/tests/test_geometric.py (hunk 4 above)
Comment (ahojnnes) [inline skimage/transform/tests/test_geometric.py]: Done.

**Comment by ahojnnes** [comment]
@jni Thanks for the review. I think, I addressed all your comments.

**Comment by jni** [comment]
Hmmm some weird import errors in Py 2.x builds, data isn't importing correctly...

**Comment by ahojnnes** [comment]
@jni There was a unicode character in one of the references. Should be fixed now.

**Comment by ahojnnes** [comment]
@jni Thanks for reviewing!

@scikit-image/core This seems to be approved by Juan, anyone has time to look over it as well?

File: doc/examples/transform/plot_fundamental_matrix.py [hunk 14]
```diff
@@ -0,0 +1,83 @@
...
+descriptor_extractor.detect_and_extract(img_left)
+keypoints_left = descriptor_extractor.keypoints
+descriptors_left = descriptor_extractor.descriptors
+
+descriptor_extractor.detect_and_extract(img_right)
+keypoints_right = descriptor_extractor.keypoints
+descriptors_right = descriptor_extractor.descriptors
+
+matches = match_descriptors(descriptors_left, descriptors_right,
+                              cross_check=True)
```
Comment (soupault) [inline doc/examples/transform/plot_fundamental_matrix.py]: Indentation

File: skimage/data/__init__.py [hunk 15]
```diff
@@ -321,3 +326,62 @@ def rocket():
...
+    in the following and the included ground-truth disparity map are valid for
+    the down-sampled images::
+
+        Focal length:           994.978px
+        Principal point x:      311.193px
+        Principal point y:      254.877px
+        Principal point dx:      31.086px
+        Baseline:               193.001mm
+
+
```
Comment (soupault) [inline skimage/data/__init__.py]: Remove one blank line

File: skimage/transform/_geometric.py [hunk 16]
```diff
@@ -208,8 +209,292 @@ def __add__(self, other):
...
+        -------
+        success : bool
+            True, if model estimation succeeds.
+
+        """
+
+        F_normalized, src_matrix, dst_matrix = \
+            self._setup_constraint_matrix(src, dst)
+
+        # Enforcing the internal constraint that two singular values must non-
```
Comment (soupault) [inline skimage/transform/_geometric.py]: *must be

Comment (ahojnnes) [inline skimage/transform/_geometric.py]: Done.

File: skimage/data/__init__.py (hunk 15 above)
Comment (ahojnnes) [inline skimage/data/__init__.py]: Done.

**Comment by ahojnnes** [comment]
@scikit-image/core Fixed the last remaining comment by @soupault. This should be ready to merge.

**Comment by soupault** [comment]
Double checked this locally, everything is good. Travis error is unrelated.

Thanks a lot @ahojnnes !
